Load also abstract types

tstrachota opened the pull request:
It's probably rare but some WSDL files use abstract types in operation definitions. Without this change it's impossible to build messages for such operations.

I'm also bumping bundler and rake dependencies here.

leejarvis commented:
> some WSDL files use abstract types in operation definitions.

Interesting. It sounds like this is breaking the spec to me (whatever that really means in the SOAP world). If a type is abstract it means that other types can derive from it, but it should never be used directly.

I think for this reason we should avoid changing the existing `types` method and perhaps consider introducing an `abstract_types` method. This means that a WSDL that does not follow the specification can still be parsed, we just have to add a little more work as consumers of the Lolsoap library. WDYT?

tstrachota commented:
> consider introducing an abstract_types method

That is one of options that I was considering. If using and abstract type is against specs then I agree with you and I'll add that method. You're right it's safer too. Thanks!

tstrachota commented:
I pushed a commit that loads abstract types into a separate hash. Unfortunately types for description of operations are handed over from `WSDL` class directly so that change itself wasn't sufficient and I had to add a flag `:allow_abstract_types`. I didn't notice it at first glance. It keeps the default behavior unchanged so risks should be limited.

Alternatively I could add a standalone class with this behavior - I have no strong preference about this.

WDYT? I'll merge it with the previous commit once it's acked.

leejarvis commented:
Thanks @tstrachota 

> Alternatively I could add a standalone class with this behavior - I have no strong preference about this.

Do you mean a new standalone class like `AbstractType` instead of `Type`? Which would remove the `allow_abstract_types ` behaviour? If so, I think that's the way to do. AFAICT the parser `type` method is only used in `Type#base_type` so having a new flag which dictates whether abstract types should be allowed or not is odd, especially in the context of calling `WSDL#abstract_types` which would return the types regardless of the value of this flag. WDYT?

tstrachota commented:
I meant rather having a standalone descendant of `WSDL` that would implement the behavior that comes with `:allow_abstract_types => true`. I might be overlooking something but I think adding a class for `AbstractType` wouldn't help us here.

The problem is `Lolsoap::Builder` checks for types and their allowed elements internally. I added `WSDL#abstract_types` just to complement `WSDL#types` but in fact only adding this method can't help with fixing the issue.

The crucial part is to allow `NamedTypeReference` to use abstract types here:
https://github.com/loco2/lolsoap/blob/master/lib/lolsoap/wsdl/named_type_reference.rb#L12
The current behaviour (without this patch) is that `WSDL#type` returns `NullType` for abstract types there and the builder raises an exception when it attempts to add an inner element.

I added the flag to make sure we stay in line with specs by default and only allow abstract types as opt in.

Does it make sense to you?

tstrachota commented:
Thanks. I squashed the commits, merging now.
